drgn_tools/rds.py

```diff
@@ -570,6 +578,55 @@ def rds_dev_info(
     else:
         return None
 
+def rdma_resource_usage(prog: Program, outfile: Optional[str] = None, report: bool = False) -> None:
+    """
+    Print RDMA restrack resource usage counts for ALL mlx5_* devices, similar to 'rdma res show'
+
+    :param prog: drgn.Program
+    :param outfile: A file to write the output to.
+    :param report: Whether to open file in append mode for report.
+    """
+    from drgn.helpers.linux import xa_for_each, list_for_each_entry
+    from drgn import container_of
+
+    dev_kset = prog["devices_kset"]
+    data = [["Index", "Device", "PD", "CQ", "QP", "CM_ID", "MR", "CTX", "SRQ"]]
+    index = 0
+
+    for dev in list_for_each_entry("struct device", dev_kset.list.address_of_(), "kobj.entry"):
+        try:
+            name = dev.kobj.name.string_().decode()
+            ib_dev = container_of(dev, "struct ib_device", "dev")
+            dev_name = ib_dev.name.string_().decode()
+            if not dev_name.startswith("mlx"):
+                continue
+            counts = {}
+            for i, res_name in RDMA_RES_TYPES.items():
```

Comment:
Rather than using `RDMA_RES_TYPES`, you can do the following. First, outside of all of the loops, look up the type:

```python
res_types_enum = prog.type("enum rdma_restrack_type")
```

Then, you can replace this loop with:

```suggestion
            for name, i in res_types_enum.enumerators:
                res_name = name[len("RDMA_RESTRACK_"):].lower()
```